### PR TITLE
Mac kext: Don't try to message offline providers, fixes missing root assert

### LINKED ISSUE #882: Mac: Kernel Panic when trying to clone repo after in-place upgrade
When testing the update flow of having an existing version of VFS4G installed, running the .pkg installer (without unmounting any previous repos) and then cloning again, triggering the kext load prompt from mount, I hit this new kernel panic. 
It's very possible that this is related to a version mismatch since we never unloaded the kext or shut down the `GVFS.Mount` process before replacing the version of PrjFSKext on disk.

```
Anonymous UUID:       2DED5D5E-E5E8-4610-D75A-4E1FA9CCE8CC

Mon Mar  4 16:46:23 2019

*** Panic Report ***
panic(cpu 2 caller 0xffffff7f9412941d): /Users/vsts/agent/2.148.0/work/1/s/ProjFS.Mac/PrjFSKext/VirtualizationRoots.cpp:659 Assertion failed: s_virtualizationRoots[rootIndex].path[0] != '\0' : When converting an absolute path to a virtualization root-relative path, the root's path should not be empty.
Backtrace (CPU 2), Frame : Return Address
0xffffffa3fe2bb440 : 0xffffff800d7aeb0d mach_kernel : _handle_debugger_trap + 0x48d
0xffffffa3fe2bb490 : 0xffffff800d8e8653 mach_kernel : _kdp_i386_trap + 0x153
0xffffffa3fe2bb4d0 : 0xffffff800d8da07a mach_kernel : _kernel_trap + 0x4fa
0xffffffa3fe2bb540 : 0xffffff800d75bca0 mach_kernel : _return_from_trap + 0xe0
0xffffffa3fe2bb560 : 0xffffff800d7ae527 mach_kernel : _panic_trap_to_debugger + 0x197
0xffffffa3fe2bb680 : 0xffffff800d7ae373 mach_kernel : _panic + 0x63
0xffffffa3fe2bb6f0 : 0xffffff7f9412941d org.vfsforgit.PrjFSKext : __Z38VirtualizationRoot_GetRootRelativePathsPKc + 0x18e
0xffffffa3fe2bb730 : 0xffffff7f9412a1f2 org.vfsforgit.PrjFSKext : __ZL32TrySendRequestAndWaitForResponses11MessageTypeP5vnodeRK9FsidInodePKciS6_PiS7_ + 0x9a
0xffffffa3fe2bb7f0 : 0xffffff7f94129c06 org.vfsforgit.PrjFSKext : __ZL21HandleFileOpOperationP5ucredPvimmmm + 0x1da
0xffffffa3fe2bb8a0 : 0xffffff800dc87d80 mach_kernel : _kauth_authorize_fileop + 0x160
0xffffffa3fe2bb900 : 0xffffff800da20b04 mach_kernel : _vn_open_auth + 0xb04
0xffffffa3fe2bb980 : 0xffffff800da0a407 mach_kernel : _open1 + 0x147
0xffffffa3fe2bbbc0 : 0xffffff800da0af44 mach_kernel : _open + 0x194
0xffffffa3fe2bbf40 : 0xffffff800ddb62bb mach_kernel : _unix_syscall64 + 0x26b
0xffffffa3fe2bbfa0 : 0xffffff800d75c466 mach_kernel : _hndl_unix_scall64 + 0x16
      Kernel Extensions in backtrace:
         org.vfsforgit.PrjFSKext(1.0)[9A47A43B-2139-38C5-9692-28F0AC67B91E]@0xffffff7f94127000->0xffffff7f9413bfff

BSD process name corresponding to current thread: GVFS.Mount
Boot args: debug=0x14c pmuflags=1 keepsyms=1

Mac OS version:
18D109

Kernel version:
Darwin Kernel Version 18.2.0: Thu Dec 20 20:46:53 PST 2018; root:xnu-4903.241.1~1/RELEASE_X86_64
Kernel UUID: 1970B070-E53F-3178-83F3-1B95FA340695
Kernel slide:     0x000000000d400000
Kernel text base: 0xffffff800d600000
__HIB  text base: 0xffffff800d500000
System model name: MacPro6,1 (Mac-F60DEB81FF30ACF6)

System uptime in nanoseconds: 365697863321079
last loaded kext at 365694351318764: org.vfsforgit.PrjFSKext	1 (addr 0xffffff7f94127000, size 86016)
loaded kexts:
org.vfsforgit.PrjFSKext	1
io.gvfs.PrjFSKext	1
com.apple.filesystems.msdosfs	1.10
com.apple.filesystems.smbfs	3.3.1
com.apple.driver.AudioAUUC	1.70
com.apple.fileutil	18.306.12
com.apple.filesystems.autofs	3.0
com.apple.driver.ApplePlatformEnabler	2.7.0d0
com.apple.driver.AGPM	110.23.46
com.apple.driver.X86PlatformShim	1.0.0
com.apple.driver.AppleUpstreamUserClient	3.6.5
com.apple.driver.AppleMCCSControl	1.5.6
com.apple.kext.AMDFramebuffer	2.0.4
com.apple.kext.AMDRadeonX4000	2.0.4
com.apple.driver.AppleGraphicsDevicePolicy	3.28.4
com.apple.AGDCPluginDisplayMetrics	3.28.4
com.apple.driver.AppleHDAHardwareConfigDriver	282.10
com.apple.driver.AppleHDA	282.10
com.apple.driver.AppleHV	1
com.apple.iokit.IOUserEthernet	1.0.1
com.apple.iokit.IOBluetoothSerialManager	6.0.10f1
com.apple.driver.pmtelemetry	1
com.apple.Dont_Steal_Mac_OS_X	7.0.0
com.apple.driver.eficheck	1
com.apple.driver.ACPI_SMC_PlatformPlugin	1.0.0
com.apple.driver.AppleLPC	3.1
com.apple.driver.AppleThunderboltIP	3.1.2
com.apple.kext.AMD7000Controller	2.0.4
com.apple.driver.AppleMGPUPowerControl	3.28.4
com.apple.driver.AppleIntelMCEReporter	107
com.apple.driver.AppleOSXWatchdog	1
com.apple.driver.AppleFIVRDriver	4.1.0
com.apple.driver.AppleIntelSlowAdaptiveClocking	4.0.0
com.apple.iokit.IOBluetoothUSBDFU	6.0.10f1
com.apple.driver.AppleFileSystemDriver	3.0.1
com.apple.driver.AppleVirtIO	2.1.3
com.apple.filesystems.hfs.kext	407.200.4
com.apple.AppleFSCompression.AppleFSCompressionTypeDataless	1.0.0d1
com.apple.BootCache	40
com.apple.AppleFSCompression.AppleFSCompressionTypeZlib	1.0.0
com.apple.AppleSystemPolicy	1.0
com.apple.filesystems.apfs	945.241.4
com.apple.driver.AirPort.BrcmNIC	1400.1.1
com.apple.iokit.AppleBCM5701Ethernet	10.3.3
com.apple.driver.AppleAHCIPort	329.200.2
com.apple.private.KextAudit	1.0
com.apple.driver.AppleACPIEC	6.1
com.apple.driver.AppleACPIButtons	6.1
com.apple.driver.AppleHPET	1.8
com.apple.driver.AppleRTC	2.0
com.apple.driver.AppleSMBIOS	2.1
com.apple.driver.AppleAPIC	1.7
com.apple.driver.AppleIntelCPUPowerManagementClient	220.0.0
com.apple.nke.applicationfirewall	190
com.apple.security.TMSafetyNet	8
com.apple.driver.AppleIntelCPUPowerManagement	220.0.0
com.apple.kext.triggers	1.0
com.apple.driver.AppleSMBusController	1.0.18d1
com.apple.kext.AMDRadeonX4030HWLibs	1.0
com.apple.iokit.IOAcceleratorFamily2	404.2.2
com.apple.kext.AMDRadeonX4000HWServices	2.0.4
com.apple.iokit.IOAVBFamily	710.1
com.apple.plugin.IOgPTPPlugin	700.7
com.apple.driver.DspFuncLib	282.10
com.apple.kext.OSvKernDSPLib	527
com.apple.driver.AppleSSE	1.0
com.apple.iokit.IOSurface	255.1
com.apple.AppleGPUWrangler	3.28.4
com.apple.driver.AppleHDAController	282.10
com.apple.iokit.IOHDAFamily	282.10
com.apple.driver.IOPlatformPluginLegacy	1.0.0
com.apple.driver.AppleSMBusPCI	1.0.14d1
com.apple.kext.AMDSupport	2.0.4
com.apple.AppleGraphicsDeviceControl	3.28.4
com.apple.driver.AppleGraphicsControl	3.28.4
com.apple.driver.AppleBacklightExpert	1.1.0
com.apple.iokit.IONDRVSupport	530
com.apple.iokit.IOGraphicsFamily	530.14
com.apple.iokit.IOSlowAdaptiveClockingFamily	1.0.0
com.apple.driver.X86PlatformPlugin	1.0.0
com.apple.driver.IOPlatformPluginFamily	6.0.0d8
com.apple.driver.AppleThunderboltEDMSink	4.2.0
com.apple.driver.AppleThunderboltPCIUpAdapter	2.1.4
com.apple.driver.AppleThunderboltDPOutAdapter	5.5.8
com.apple.iokit.BroadcomBluetoothHostControllerUSBTransport	6.0.10f1
com.apple.iokit.IOBluetoothHostControllerUSBTransport	6.0.10f1
com.apple.iokit.IOBluetoothHostControllerTransport	6.0.10f1
com.apple.iokit.IOBluetoothFamily	6.0.10f1
com.apple.driver.AppleUSBAudio	315.6
com.apple.driver.usb.IOUSBHostHIDDevice	1.2
com.apple.iokit.IOAudioFamily	206.5
com.apple.vecLib.kext	1.2.0
com.apple.driver.usb.cdc	5.0.0
com.apple.driver.usb.networking	5.0.0
com.apple.driver.usb.AppleUSBHostCompositeDevice	1.2
com.apple.driver.usb.AppleUSBHub	1.2
com.apple.iokit.IOSerialFamily	11
com.apple.filesystems.hfs.encodings.kext	1
com.apple.driver.AppleThunderboltDPInAdapter	5.5.8
com.apple.driver.AppleThunderboltDPAdapterFamily	5.5.8
com.apple.driver.AppleThunderboltPCIDownAdapter	2.1.4
com.apple.driver.AppleXsanScheme	3
com.apple.driver.AppleThunderboltNHI	4.7.6
com.apple.iokit.IOThunderboltFamily	6.8.1
com.apple.iokit.IOAHCIBlockStorage	301.200.2
com.apple.driver.usb.AppleUSBXHCIPCI	1.2
com.apple.driver.usb.AppleUSBXHCI	1.2
com.apple.driver.AppleUSBMergeNub	900.4.2
com.apple.iokit.IO80211Family	1200.12.2
com.apple.driver.corecapture	1.0.4
com.apple.iokit.IOEthernetAVBController	1.1.0
com.apple.driver.mDNSOffloadUserClient	1.0.1b8
com.apple.driver.usb.AppleUSBEHCIPCI	1.2
com.apple.driver.usb.AppleUSBEHCI	1.2
com.apple.iokit.IOAHCIFamily	288
com.apple.driver.usb.AppleUSBHostPacketFilter	1.0
com.apple.iokit.IOUSBFamily	900.4.2
com.apple.driver.AppleUSBHostMergeProperties	1.2
com.apple.iokit.IOSMBusFamily	1.1
com.apple.driver.AppleEFINVRAM	2.1
com.apple.iokit.IOHIDFamily	2.0.0
com.apple.driver.AppleEFIRuntime	2.1
com.apple.security.quarantine	3
com.apple.security.sandbox	300.0
com.apple.kext.AppleMatch	1.0.0d1
com.apple.driver.DiskImages	493.0.0
com.apple.driver.AppleFDEKeyStore	28.30
com.apple.driver.AppleEffaceableStorage	1.0
com.apple.driver.AppleKeyStore	2
com.apple.driver.AppleUSBTDM	456.230.1
com.apple.driver.AppleMobileFileIntegrity	1.0.5
com.apple.kext.CoreTrust	1
com.apple.iokit.IOUSBMassStorageDriver	145.200.2
com.apple.iokit.IOSCSIBlockCommandsDevice	408.200.1
com.apple.iokit.IOSCSIArchitectureModelFamily	408.200.1
com.apple.iokit.IOStorageFamily	2.1
com.apple.driver.AppleCredentialManager	1.0
com.apple.driver.KernelRelayHost	1
com.apple.iokit.IOUSBHostFamily	1.2
com.apple.driver.usb.AppleUSBCommon	1.0
com.apple.driver.AppleBusPowerController	1.0
com.apple.driver.AppleSEPManager	1.0.1
com.apple.driver.IOSlaveProcessor	1
com.apple.iokit.IOReportFamily	47
com.apple.iokit.IOTimeSyncFamily	700.7
com.apple.iokit.IONetworkingFamily	3.4
com.apple.driver.AppleACPIPlatform	6.1
com.apple.driver.AppleSMC	3.1.9
com.apple.iokit.IOPCIFamily	2.9
com.apple.iokit.IOACPIFamily	1.4
com.apple.kec.pthread	1
com.apple.kec.Libm	1
com.apple.kec.corecrypto	1.0

EOF
```

### PR DESCRIPTION
**Pull request by pmj**
This fixes #882.

1. The first commit adds an assertion that a root's path is not poisoned when attempting to use it. This means the bug will also be caught in debug builds.
2. The second commit fixes the actual bug - attempting to send a message to a root that doesn't have a provider from the fileop handler. This is done by checking if the root has an online provider in `ShouldHandleFileOpEvent()`.
3. The third commit fixes some kextlog messages which I noticed while debugging this.

### CODE REVIEW

File: ProjFS.Mac/PrjFSKext/KauthHandler.cpp [hunk 1]
```diff
@@ -938,6 +938,16 @@ static bool ShouldHandleFileOpEvent(
         }
     }
     
+    {
+        PerfSample checkRootOnlineSample(perfTracer, PrjFSPerfCounter_FileOp_ShouldHandle_CheckRootOnline);
+        
+        if (!VirtualizationRoot_IsOnline(*root))
```
Comment (jeschu1) [inline ProjFS.Mac/PrjFSKext/KauthHandler.cpp]: This seems like similar to the check done in HandleVnodeOperation -> TryGetVirtualizationRoot -> !VirtualizationRoot_IsOnline
HandleFileOpOperation -> ShouldHandleFileOpEvent -> !VirtualizationRoot_IsOnline

In ShouldHandleFileOpEvent, we do a VirtualizationRoot_IsValidRootHandle check but not in TryGetVirtualizationRoot.  Can you elaborate on why it's needed in one and not the othere?

**Comment by jeschu1** [comment]
Change makes sense just questions on consistency for *root

File: ProjFS.Mac/PrjFSKext/KauthHandler.cpp (hunk 1 above)
Comment (pmj) [reply ProjFS.Mac/PrjFSKext/KauthHandler.cpp]: I guess `TryGetVirtualizationRoot()` checks the 2 distinct cases of invalid handles, `RootHandle_ProviderTemporaryDirectory` and `RootHandle_None` separately.

I generally have a bunch of issues with this whole area of code:

 * `TryGetVirtualizationRoot` is terribly named for such a "kitchen sink" function.
 * There's a lot of duplication between the fileop and vnode listener checks, but no code reuse.
 * The functions take far too many parameters.

These aren't issues I'm planning to fix in this particular PR; I'd love to fix them once the fileop and vnode handlers are unit tested though.

Comment (jeschu1) [reply ProjFS.Mac/PrjFSKext/KauthHandler.cpp]: @pmj Yes Yes Yes :-)

> There's a lot of duplication between the fileop and vnode listener checks, but no code reuse.
Exactly and as you called out the checks are different, probably contributing to this bug happening.

> These aren't issues I'm planning to fix in this particular PR; I'd love to fix them once the fileop and vnode handlers are unit tested though.
Let's make that a plan!

Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/KauthHandler.cpp]: > Exactly and as you called out the checks are different, probably contributing to this bug happening.

I agree on this point as well, if we can get those code paths sharing code again that would be great 😄

File: ProjFS.Mac/PrjFSKext/public/PrjFSPerfCounter.h [hunk 2]
```diff
@@ -36,6 +36,8 @@ enum PrjFSPerfCounter : int32_t
                 PrjFSPerfCounter_FileOp_FindRoot,
                     PrjFSPerfCounter_FileOp_FindRoot_Iteration,
                 PrjFSPerfCounter_FileOp_ShouldHandle_NoRootFound,
+            PrjFSPerfCounter_FileOp_ShouldHandle_CheckRootOnline,
+                PrjFSPerfCounter_FileOp_ShouldHandle_OfflineRoot,
```
Comment (wilbaker) [inline ProjFS.Mac/PrjFSKext/public/PrjFSPerfCounter.h]: The `PerfCounterNames` array in kext-perf-traciing.cpp needs to be updated to include these new values as well.

Comment (pmj) [reply ProjFS.Mac/PrjFSKext/public/PrjFSPerfCounter.h]: Thanks for catching that, I'll fix that up shortly.

Comment (pmj) [reply ProjFS.Mac/PrjFSKext/public/PrjFSPerfCounter.h]: Fixed. And so that I can't make that mistake again, posted #887.

**Pull request merged**